pkg/sentry/socket/plugin/stack/socket.go

```diff
@@ -0,0 +1,886 @@
+	} else if ret < 0 && (-ret) != errno.EINPROGRESS {
+		/* EALREADY, EISCONN, ECONNREFUSED, EINVAL */
+		return int2err(ret)
+	}
+
+	if err := s.waitEvent(t, waiter.EventOut); err != nil {
+		return err2syserr(err)
+	}
+
+	// Call connect() again after blocking to find connect's result.
+	ret = cgo.Connect(s.fd, sockaddr)
+	if ret == 0 {
+		return nil
+	} else if ret < 0 && (-ret) == errno.EISCONN {
+		return nil
+	}
+	return int2err(ret)
+}
+
+// Shutdown implements socket.Socket.Shutdown.
+func (s *socketOperations) Shutdown(t *kernel.Task, how int) *syserr.Error {
+	ret := cgo.Shutdown(s.fd, how)
+	return int2err(ret)
+}
+
+// GetSockOpt implements socket.Socket.GetSockOpt.
+func (s *socketOperations) GetSockOpt(t *kernel.Task, level int, name int, outPtr hostarch.Addr, outLen int) (marshal.Marshallable, *syserr.Error) {
+	var optVal = make([]byte, outLen)
+	if outLen == 0 {
+		optPtr := primitive.ByteSlice(optVal)
+		return &optPtr, nil
+	}
+	rc, outLen := cgo.Getsockopt(s.fd, level, name, optVal, outLen)
+	if rc < 0 {
+		return nil, int2err(rc)
+	}
+	optPtr := primitive.ByteSlice(optVal[:outLen])
+	return &optPtr, nil
+}
+
+// SetSockOpt implements socket.Socket.SetSockOpt.
+func (s *socketOperations) SetSockOpt(t *kernel.Task, level int, name int, optVal []byte) *syserr.Error {
+	rc := cgo.Setsockopt(s.fd, level, name, optVal)
+	return int2err(rc)
+}
+
+// State implements socket.Socket.State.
+func (s *socketOperations) State() uint32 {
+	var optVal = make([]byte, 1)
+	rc, _ := cgo.Getsockopt(s.fd, syscall.SOL_TCP, syscall.TCP_INFO, optVal, 1)
+	if rc < 0 {
+		return 0
+	}
+	return uint32(optVal[0])
+}
+
+// Type implements socket.Socket.Type.
+func (s *socketOperations) Type() (family int, skType linux.SockType, protocol int) {
+	return s.family, s.skType, s.protocol
+}
+
+// OnClose implements vfs.FileDescriptionImpl.OnClose.
+func (s *socketOperations) OnClose(ctx context.Context) error {
+	return nil
+}
+
+// EventRegister implements waiter.Waitable.EventRegister.
+func (s *socketOperations) EventRegister(e *waiter.Entry) error {
+	s.eventInfo.Wq.EventRegister(e)
+	notifier.UpdateFD(s.fd)
+	return nil
+}
+
+// EventUnregister implements waiter.Waitable.EventUnregister.
+func (s *socketOperations) EventUnregister(e *waiter.Entry) {
+	s.eventInfo.Wq.EventUnregister(e)
+	notifier.UpdateFD(s.fd)
+}
+
+// Readiness implements socket.Socket.Readiness.
+func (s *socketOperations) Readiness(mask waiter.EventMask) waiter.EventMask {
+	var events waiter.EventMask
+
+	evInfo := &s.eventInfo
+	iomask := mask & (waiter.EventIn | waiter.EventOut)
+	// If EventIn and EventOut are both registered, and only one event is reported,
+	// fd_ready is needed to assure the other event is not missed.
+	if evInfo.Ready&mask == 0 || evInfo.Ready&iomask != iomask {
+		events = waiter.EventMask(cgo.Readiness(s.fd, uint64(mask)))
+	} else {
+		events = evInfo.Ready & mask
+		evInfo.Ready &= ^(mask & (waiter.EventIn | waiter.EventOut | waiter.EventHUp))
+	}
```

Comment:
Sorry for the confusion here. We have refactored this part to make it more readable hopefully.
Basically, there are two conditions: fast path and slow path for readiness.
If the requested event is already cached in sentry (i.e. eventInfo.Ready field of the socket struct), which is done by the I/O thread of plugin stack, we can get the ready events without invoking plugin stack's Readiness again. Before returning, we clean-up the event cache since we have consumed the events already. This will be regarded as the fast path.
Otherwise, we fall back to the slow path and let plugin stack check whether the event is ready.